### ISSUE TITLE
[LocalFunction] Shape mismatch attempting to re-use buffer

### ISSUE THREAD

**Issue by BowenBao** [microsoft/onnxruntime#17061]
Similar to #16813 (unsure if related), this issue occurs when running dynamo exporter produced model w/ local functions. The fully inlined model runs successfully.

**NOTE that ORT must be built with ONNX v1.14.1 or above**, otherwise segfault may occur during ONNX shape inference.

Example model and repro script can be found here: [torchbench_hf_Bart](https://microsoft-my.sharepoint.com/:f:/p/bowbao/EteZtZ0Rw6BBqi9zrHMJPEYBdpJ54BW81XJs-4SpfhXOJA?e=sSPpuV).

> ========== hf_Bart failed: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running If node. Name:'_inline_aten_tn5' Status Message: /bert_ort/bowbao/repos/bench/onnxruntime/onnxruntime/core/framework/op_kernel.cc:83 virtual OrtValue* onnxruntime::OpKernelContext::OutputMLValue(int, const onnxruntime::TensorShape&) status.IsOK() was false. Shape mismatch attempting to re-use buffer. {768,3072} != {768,50265}. Validate usage of dim_value (values should be > 0) and dim_param (all values with the same string should equate to the same size) in shapes in the model.

> ========== inlined hf_Bart passed

**Comment by snnn** [comment]
Most likely it is an ONNX shape inference bug. Does ONNX have any unit test for this shape inference function?

**Comment by BowenBao** [comment]
This is new issue, the repro was from a local build of ORT w/ ONNX 1.14.1.

**Comment by BowenBao** [comment]
> Most likely it is an ONNX shape inference bug.

Model does pass below shape inference checks without error, this is part of the repro provided.

```python
    onnx.shape_inference.infer_shapes(
        onnx_model, check_type=True, strict_mode=True, data_prop=True
    )
    onnx.checker.check_model(onnx_model, full_check=True)
```

**Comment by snnn** [comment]
It doesn't mean the shape inference functions generate correct results. It didn't crash, but the results could be wrong. For example, if a matmul has two inputs: matrix A and matrix B.  A has shape: [m,n]. B has shape [n,k], then the result should have shape [m,k]. But the shape inference functions wrongly thinks the output shape is [m,m], onnx checker would still pass. However, ONNX Runtime would allocate a wrong buffer for the output.  
So, what I want to say is: ONNX should have a way to test if the shape inference functions are correct, and it shouldn't rely on ORT. Otherwise it would be like: "I must be right because I say I am right".  If the correctness of the shape inference functions is based on ORT, then ORT cannot trust the function's results.

**Comment by yuslepukhin** [comment]
I will take a look at it

**Comment by jcwchen** [comment]
> Model does pass below shape inference checks without error, this is part of the repro provided.

ONNX shape inference might overlook some issues coming from subgraph or local function due to https://github.com/onnx/onnx/issues/5463. Fresh PR https://github.com/onnx/onnx/pull/5488 should fix that and we can run shape inference again with that fix (and enable strict_mode) to ensure whether ONNX can catch it.

**Comment by yuslepukhin** [comment]
@jcwchen Is this going to be in 1.14.1?

**Comment by jcwchen** [comment]
> @jcwchen Is this going to be in 1.14.1?

No, because https://github.com/onnx/onnx/pull/5488 hasn't been merged yet and ONNX 1.14.1 will probably be out this week. 

To workaround, if you want onnx shape inference be able to catch local function related issues, you can try https://pypi.org/project/onnx-weekly/ instead after https://github.com/onnx/onnx/pull/5488 has been merged.

**Comment by yuslepukhin** [comment]
I am actually debugging in C++ with ORT built with ONNX 1.14.1

**Comment by jcwchen** [comment]
> I am actually debugging in C++ with ORT built with ONNX 1.14.1

IIUC, this ONNX issue (not honor check_type and strict_mode for local function) should be fine for ORT, because ORT does explicitly set check_type and strict_mode when using shape inference for location function:

https://github.com/microsoft/onnxruntime/blob/289600b47dfcf837365d672a3a69ed823f989163/onnxruntime/core/graph/function_utils.cc#L333

The fix https://github.com/onnx/onnx/pull/5488 is only solving the issue that ONNX's shape inference API suppresses local function related shape inference error.

**Comment by yuslepukhin** [comment]
The inlined model runs and issues the below error. I think the model needs to be checked.

> 2023-08-14 14:55:24.0335656 [E:onnxruntime:, sequential_executor.cc:514 onnxruntime::ExecuteKernel] Non-zero status code returned while running Gather node. Name:'n0__6' Status Message: indices element out of data bounds, idx=-1992083005621533730 must be within the inclusive range [-50265,50264]
Exception: Non-zero status code returned while running Gather node. Name:'n0__6' Status Message: indices element out of data bounds, idx=-1992083005621533730 must be within the inclusive range [-50265,50264]

**Comment by yuslepukhin** [comment]
The inlined version of the model runs w/o any errors. Most of the time is spent on optimizations when enabled.

**Comment by jcwchen** [comment]
> The inlined version of the model runs w/o any errors. Most of the time is spent on optimizations when enabled.

Sounds like it should throw some errors, but it doesn't? If yes, there might be other unidentified issues/limitations for ONNX's function shape inference and https://github.com/onnx/onnx/pull/5488 won't fix that I think.

**Comment by yuslepukhin** [comment]
The non-inlined version does re-pro the error in the description. The inlining is done using onnx. The  script also performes shape inferencing. I will try to run this without it and see what happens.

**Comment by yuslepukhin** [comment]
The root cause of the problem has been identified. We are going to issue a temporary fix.
Here is the known workarounds meanwhile.

- You can inline and pre-optimize the model and save it. No need for ONNX inlining. The built-in inlining works just fine.  `session_options.optimized_model_filepath = optimized_model_name.onnx`. You can then load the optimized model fast and run it. (Recommended option)
- Disable memory reuse `session_options.enable_mem_reuse = False`. This would increase memory consumption and would likely impact inference performance.

**Comment by jberclaz** [comment]
@yuslepukhin , can you comment on the fix for this issue? Has it been included in release 1.16.0? Thanks

**Comment by yuslepukhin** [comment]
> @yuslepukhin , can you comment on the fix for this issue? Has it been included in release 1.16.0? Thanks

I believe so. It was approved for that.
